File: src/main/java/mainPackage/CustomTableRow.java
package mainPackage;

import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.collections.ObservableList;
import javafx.scene.control.*;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseButton;
import javafx.util.Duration;

import java.lang.reflect.Field;

/**
 * Created by jacobmenke on 4/23/17.
 */
public class CustomTableRow extends TableRow<FileInfo> {

    public static void changeToolTipTime(Tooltip tooltip, Integer time) {
        try {
            Field fieldBehavior = tooltip.getClass().getDeclaredField("BEHAVIOR");

            fieldBehavior.setAccessible(true);
            Object objBehavior = fieldBehavior.get(tooltip);


            Field fieldTimer = objBehavior.getClass().getDeclaredField("activationTimer");
            fieldTimer.setAccessible(true);
            Timeline objTimer = (Timeline) fieldTimer.get(objBehavior);

            objTimer.getKeyFrames().clear();
            objTimer.getKeyFrames().add(new KeyFrame(new Duration(time*1000)));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    private Tooltip tooltip = new Tooltip();


    public CustomTableRow(TableView mainTableView, ObservableList<FileInfo> files, MainController mainController) {
        CommonUtilities.formatTooltip(tooltip);


        this.setOnMouseClicked(e-> {
            if (e.getButton() == MouseButton.SECONDARY) {
                FileInfo fileInfo = (FileInfo)this.getItem();


                ContextMenu cm = Utilities.createContextMenu(fileInfo, mainController.mainTableView, mainController.files, mainController, "tableView");
                cm.show(this, e.getScreenX(), e.getScreenY());
            }
        });
    }



    @Override
    protected void updateItem(FileInfo item, boolean empty) {
        super.updateItem(item, empty);

        if (item == null){
            setTooltip(null);
        } else {
            FileInfo fileInfo = item;

            changeToolTipTime(tooltip,3);
            tooltip.setText(fileInfo.toString());

            if (!item.isDirectory()){
                tooltip.setGraphic(new ImageView(fileInfo.getFileImage()));
            } else {

                if (FilePathTreeItem.specialDirs.containsKey(item.getAbsolutePath())) {
                    tooltip.setGraphic(new ImageView(FilePathTreeItem.specialDirs.get(item.getAbsolutePath())));

                } else {
                    tooltip.setGraphic(new ImageView(FilePathTreeItem.folderCollapseImage));
                }

            }

            setTooltip(tooltip);
        }

    }
}

File: src/main/kotlin/mainPackage/CustomTableRow.kt
package mainPackage

import javafx.collections.ObservableList
import javafx.scene.control.ContextMenu
import javafx.scene.control.TableRow
import javafx.scene.control.TableView
import javafx.scene.control.Tooltip
import javafx.scene.image.ImageView
import javafx.scene.input.MouseButton

class CustomTableRow(
    mainTableView: TableView<FileInfo>,
    files: ObservableList<FileInfo>,
    mainController: MainController
) : TableRow<FileInfo>() {
    var cm: ContextMenu? = null
    private val tooltip = Tooltip()

    init {
        CommonUtilities.formatTooltip(tooltip)
        styleProperty().bind(CommonUtilities.tableViewColorProperty)
        setOnMouseClicked { e ->
            if (e.button == MouseButton.SECONDARY) {
                val fileInfo = item
                cm?.hide()
                cm = Utilities.createContextMenu(fileInfo, mainController.mainTableView, mainController.files, mainController, "tableView")
                cm?.show(this, e.screenX, e.screenY)
            }
        }
    }

    override fun computePrefHeight(width: Double): Double = font.size * 2.5

    override fun updateItem(item: FileInfo?, empty: Boolean) {
        super.updateItem(item, empty)
        if (item == null) {
            setTooltip(null)
        } else {
            changeToolTipTime(tooltip, 3)
            tooltip.text = item.toString()
            if (!item.isDirectory) {
                tooltip.graphic = ImageView(item.getFileImage())
            } else {
                val img = FilePathTreeItem.specialDirs[item.absolutePath]
                tooltip.graphic = ImageView(img ?: FilePathTreeItem.folderCollapseImage)
            }
            setTooltip(tooltip)
        }
    }

    companion object {
        @JvmStatic
        fun changeToolTipTime(tooltip: Tooltip, time: Int) {
            tooltip.showDelay = javafx.util.Duration.seconds(time.toDouble())
        }
    }
}
